features/remotepollerng/daemon/src/main/java/org/opennms/netmgt/remotepollerng/PerspectiveServiceTracker.java

```diff
@@ -0,0 +1,259 @@
+            return Objects.hash(this.nodeId, this.ipAddress, this.serviceName, this.perspectiveLocation);
+        }
+
+        public static ServicePerspectiveRef from(final ServicePerspective servicePerspective) {
+            return new ServicePerspectiveRef(servicePerspective.getService().getNodeId(),
+                                             servicePerspective.getService().getIpAddress(),
+                                             servicePerspective.getService().getServiceName(),
+                                             servicePerspective.getPerspectiveLocation().getLocationName());
+        }
+    }
+
+    private final SessionUtils sessionUtils;
+
+    private final ApplicationDao applicationDao;
+
+    private final Set<Session> sessions = Sets.newHashSet();
+
+    private final Timer timer = new Timer("PerspectiveServiceTracker-Timer");
+
+    @Autowired
+    public PerspectiveServiceTracker(final SessionUtils sessionUtils,
+                                     final ApplicationDao applicationDao) {
+        this.sessionUtils = Objects.requireNonNull(sessionUtils);
+        this.applicationDao = Objects.requireNonNull(applicationDao);
+
+        final long timerIntevalMs = Math.min(REFRESH_RATE_LIMIT_MS, TimeUnit.SECONDS.toMillis(5));
+        this.timer.schedule(new TimerTask() {
+            @Override
+            public void run() {
+                PerspectiveServiceTracker.this.update(false);
+            }
+        }, timerIntevalMs, timerIntevalMs);
+    }
+
+    public AutoCloseable track(final Listener listener) {
+        final Session session = new Session(listener);
+        session.update(true);
+
+        return session;
+    }
+
+    @Override
+    public void destroy() throws Exception {
+        this.timer.cancel();
+    }
+
+    @EventHandler(ueis = {
+            EventConstants.NODE_GAINED_SERVICE_EVENT_UEI,
+            EventConstants.SERVICE_DELETED_EVENT_UEI,
+            EventConstants.NODE_CATEGORY_MEMBERSHIP_CHANGED_EVENT_UEI,
+            EventConstants.NODE_LOCATION_CHANGED_EVENT_UEI,
+            EventConstants.NODE_ADDED_EVENT_UEI,
+            EventConstants.NODE_DELETED_EVENT_UEI,
+            EventConstants.NODE_GAINED_INTERFACE_EVENT_UEI,
+            EventConstants.INTERFACE_DELETED_EVENT_UEI,
+            EventConstants.INTERFACE_REPARENTED_EVENT_UEI,
+            EventConstants.SUSPEND_POLLING_SERVICE_EVENT_UEI,
+            EventConstants.RESUME_POLLING_SERVICE_EVENT_UEI,
+            EventConstants.APPLICATION_CHANGED_EVENT_UEI,
+            EventConstants.APPLICATION_CREATED_EVENT_UEI,
+            EventConstants.APPLICATION_DELETED_EVENT_UEI,
+    })
+    public void handleEvent(final IEvent event) {
+        update(true);
+    }
+
+    private void update(final boolean dirty) {
+        synchronized (PerspectiveServiceTracker.this.sessions) {
+            try {
+                PerspectiveServiceTracker.this.sessions.forEach(session -> session.update(dirty));
```

Comment:
Every session will create its own transaction and repeat the same DB query. This is not a big problem since there's likely only 1 session, but would be nicer to do that outside the loop.

Reply:
But the session is only created if an update is required for that session. By moving it out of the session, it will be created every time. Most of the time, calls to the `update` method will result in an early exit.